tests/core/gadget-kernel-refs-update-pc/task.yaml

```diff
@@ -0,0 +1,129 @@
+    PC_SNAP_ID: UqFziVZDHLSyO3TqSWgNBoAdHbLI4dAH
+    # snap-id of 'pc-kernel' snap
+    PC_KERNEL_ID: pYVQrBcKmBa0mZ4CCN7ExT6jH8rY1hza
+    START_REVISION: 1000
+
+# the test is only meaningful on core devices
+# XXX: uc20
+systems: [ubuntu-core-1*]
+
+prepare: |
+    if [ "$TRUST_TEST_KEYS" = "false" ]; then
+        echo "This test needs test keys to be trusted"
+        exit
+    fi
+    snap ack "$TESTSLIB/assertions/testrootorg-store.account-key"
+    #shellcheck source=tests/lib/store.sh
+    . "$TESTSLIB"/store.sh
+    setup_fake_store "$BLOB_DIR"
+    readlink /snap/pc/current > pc-revision.txt
+    readlink /snap/pc-kernel/current > pc-kernel-revision.txt
+    # create kernel-ref kernel/gadget pair
+    cp /var/lib/snapd/snaps/pc-kernel_*.snap kernel.snap
+    unsquashfs -d pc-kernel-snap kernel.snap
+    # prepare kernel with content
+    mkdir pc-kernel-snap/some-dir/
+    touch pc-kernel-snap/some-dir/foo-from-some-dir
+    touch pc-kernel-snap/some-dir/some-subdir
+    touch pc-kernel-snap/some-file
+    cat >> pc-kernel-snap/meta/kernel.yaml<<EOF
+    assets:
+      test:
+        update: true
+        content:
+          - some-dir/
+          - some-file
+    EOF
+    snap pack --filename=pc-kernel_x1.snap pc-kernel-snap
+    # has to be asserted
+    cat <<EOF > decl-headers.json
+    {"snap-id": "$PC_KERNEL_ID"}
+    EOF
+    cat <<EOF > rev-headers.json
+    {"snap-id": "$PC_KERNEL_ID", "snap-revision": "$START_REVISION"}
+    EOF
+    new_snap_declaration "$BLOB_DIR" pc-kernel_x1.snap --snap-decl-json decl-headers.json
+    new_snap_revision "$BLOB_DIR" pc-kernel_x1.snap --snap-rev-json rev-headers.json
+    # new gadget
+    cp /var/lib/snapd/snaps/pc_*.snap gadget.snap
+    unsquashfs -d pc-snap gadget.snap
+    # prepare version with kernel ref
+    cat >> pc-snap/meta/gadget.yaml <<'EOF'
+              - source: $kernel:test/some-dir/
+                target: /
+              - source: $kernel:test/some-file
+                target: /
+    EOF
+    snap pack --filename=pc_x1.snap pc-snap
+    # has to be asserted
+    cat <<EOF > decl-headers.json
+    {"snap-id": "$PC_SNAP_ID"}
+    EOF
+    cat <<EOF > rev-headers.json
+    {"snap-id": "$PC_SNAP_ID", "snap-revision": "$START_REVISION"}
+    EOF
+    new_snap_declaration "$BLOB_DIR" pc_x1.snap --snap-decl-json decl-headers.json
+    new_snap_revision "$BLOB_DIR" pc_x1.snap --snap-rev-json rev-headers.json
+restore: |
+    if [ "$TRUST_TEST_KEYS" = "false" ]; then
+        echo "This test needs test keys to be trusted"
+        exit
+    fi
+    #shellcheck source=tests/lib/store.sh
+    . "$TESTSLIB"/store.sh
+    teardown_fake_store "$BLOB_DIR"
+    # XXX: ideall we would restore the kernel/gadget here but the kernel
+    #      restore requires a reboot :/
+    
+execute: |
+    if [ "$SPREAD_REBOOT" = 0 ]; then
+        # first install the asserted kernel so that the gadget can resolve
+        # the kernel refs
+        snap install pc-kernel_x1.snap
```

Comment:
notice that this order won't work  according to the plan we discussed, the gadget needs to be installed first. Which means there is probably something that needs changing in this PR or to be done in an immediate follow up

Reply:
Indeed, I added a XXX that mentions that this will change. My plan was to do this in a followup but I will also explore a bit how much work (and extra diff) it is to do it in this PR. If it's not too much diff it would make sense to have it all in a single PR.

Reply:
This is now properly addressed in the followup #9981